publisher/publisher.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package publisher
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type LogDescription struct {
+	ID        []byte
+	URI       string
+	PublicKey *ecdsa.PublicKey
+}
+
+type rawLogDescription struct {
+	URI       string `json:"uri"`
+	PublicKey string `json:"key"`
+}
+
+func (logDesc *LogDescription) UnmarshalJSON(data []byte) error {
+	var rawLogDesc rawLogDescription
+	if err := json.Unmarshal(data, &rawLogDesc); err != nil {
+		return fmt.Errorf("Failed to unmarshal log description, %s", err)
+	}
+	logDesc.URI = rawLogDesc.URI
+	// Load Key
+	pkBytes, err := base64.StdEncoding.DecodeString(rawLogDesc.PublicKey)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	pk, err := x509.ParsePKIXPublicKey(pkBytes)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	switch k := pk.(type) {
```

Comment:
It seems like ParsePKIXPublicKey should consistently return one of these two types, right? Why the switch/case?